fl4health/clients/moon_client.py

```diff
@@ -0,0 +1,107 @@
+    i.e., conducting contrastive learning in model-level.
+    """
+
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.initial_tensors: List[torch.Tensor]
+        self.contrastive_weight: float = 10
+        self.temprature: float = 0.5
+        self.current_loss: float
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
+        self.len_old_models_buffer: int = 1
+        self.old_models_list: list[MoonModel] = []
+        self.global_model: MoonModel
+        self.features: torch.Tensor
+        self.old_features_list: list[torch.Tensor] = []
+        self.global_features: torch.Tensor
+
+    def predict(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+        preds, self.features, _ = self.model(input)
+        self.features = self.features.view(len(self.features), -1)
+        self.old_features_list = []
+        for old_model in self.old_models_list:
+            _, old_features, _ = old_model(input)
+            old_features = old_features.view(len(old_features), -1).detach()
+            self.old_features_list.append(old_features)
+        _, self.global_features, _ = self.global_model(input)
+        self.global_features = self.global_features.view(len(self.global_features), -1).detach()
+        if isinstance(preds, dict):
+            return preds
+        elif isinstance(preds, torch.Tensor):
+            return {"prediction": preds}
+        else:
+            raise ValueError("Model forward did not return a tensor or dictionary or tensors")
+
+    def get_contrastive_loss(self) -> torch.Tensor:
+        assert len(self.features) == len(self.global_features)
+        posi = self.cos_sim(self.features, self.global_features)
+        logits = posi.reshape(-1, 1)
+        for old_features in self.old_features_list:
+            assert len(self.features) == len(old_features)
+            nega = self.cos_sim(self.features, old_features)
+            logits = torch.cat((logits, nega.reshape(-1, 1)), dim=1)
+        logits /= self.temprature
+        labels = torch.zeros(self.features.size(0)).to(self.device).long()
+
+        return self.ce_criterion(logits, labels)
+
+    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+        assert isinstance(self.model, MoonModel)
+
+        # Save the parameters of the old local model
+        old_model = copy.deepcopy(self.model)
```

Comment:
Probably also want to set params.requires_grad = False for all old_models params since we are not optimizing.